Effort to create a Baby Mailgun

amatai opened the pull request.

Commit:
Initialize the repo with setup requirements

Commit:
Initial step for api, store, config

Commit:
Fix store init, added dependency in setup

Commit:
Cleanup setup.py, initialize unittests

Commit:
Add message to DB + flake8 changes

Commit:
Celery hookup + smtplib 1st attemp

celery needs pymongo 2.6.3, pin it.
bson docs can't have '.' in them, replace with '_'

Commit:
Updated README

Commit:
Initial smtp sending, mx record lookup

Commit:
Working smtplib, GET call

- Implemented GET functionality
- Cleanup the sendmail task to handle DNS, SMTP errors and set status
  properly.
- Moved doc sanitizing required for bson to store
- Remove unncessary config

Commit:
Fix exception handling for smtp converstaion

Commit:
Updating README with API calls

Commit:
Added unit-tests

Unit tests for API and Store
Updated config for retry after 10 mins

Commit:
Update README - unittest and debug info

anton-efimenko commented on minimailgun/api.py:
```diff
@@ -0,0 +1,96 @@
+        JSON_AS_ASCII=False,):
+    app = Flask(app_name)
+    if debug is None:
+        debug = bool(os.environ.get('MINI_MAILGUN_DEBUG'))
```
Is there any particular reason you decided to store it in the environment rather than config file?

amatai replied:
Flask debug seemed to me a more of a dev time flag for auto-reload etc, hence didn't make sense to do in config.

anton-efimenko replied:
Makes sense!

anton-efimenko commented on minimailgun/api.py:
```diff
@@ -0,0 +1,96 @@
+    abort(500, str(exc))
+
+
+def make_json_error(exc):
```
Say you want to run baby mailgun on a production server. What would you change in this function?

amatai replied:
I'm not sure, I am seeing the issue here :(

anton-efimenko replied:
No problem! Generally you don't want the API users to see the internal errors, stacktraces of your application in the API response. It is more secure to just return "Internal server error" and log the exception.

amatai replied:
I believe, as written it will not send stacktrace in the API response. Stack-trace just being logged using logger. Am I missing something?

anton-efimenko replied:
It won't send the stack trace but it will send the error, is it correct? Honestly I haven't tested it.

amatai replied:
Yes, i agree with that.

anton-efimenko commented on minimailgun/store.py:
```diff
@@ -0,0 +1,118 @@
+        }
+
+        # add the message to db, with a unique UUID
+        # its going to be rare that we'll have UUID collision
```
It's going to be super rare: https://en.wikipedia.org/wiki/Universally_unique_identifier#Random_UUID_probability_of_duplicates

amatai replied:
Agreed! I guess, my paranoia kicked in at that point.

anton-efimenko commented on minimailgun/store.py:
```diff
@@ -0,0 +1,118 @@
+        for _ in range(2):
+            mail['_id'] = uuid.uuid4()
+            try:
+                result = self.db.mails.insert(mail, j=True)
```
MongoDB provides you with an ID when you insert a document: http://docs.mongodb.org/v3.0/reference/method/db.collection.insert/#id-field

Is there any particular reason you didn't use it?

amatai replied:
Yes, I wanted the id, to be a field that I could return back in the response and use that as the resource identifier as well. Mongo's ObjectId is a 12-byte number, while UUID is 16-byte. UUID is more prevalent in resource identification, hence a lot of code exists to handle that both server and client side. As an example, werkzeug provides a converter for string uuid to uuid.UUID object

Also, mongo makes _id unique by default, I didn't have to create another index to ensure the UUID I chose would be unique.

amatai replied:
I got curious, why Mongo had chosen ObjectID instead of uuid to make the _id field. If there were any benefits? And, I ran across this nugget: https://jira.mongodb.org/browse/JAVA-403

There's an attachment showing the performance of insert for different types of _id (objectId, uuid). Had I researched this before, I would probably go with ObjectId and not go by convenience (in the way of utility methods) of uuid. You learn something everyday.

anton-efimenko replied:
And that's great! :)

anton-efimenko commented on minimailgun/tasks.py:
```diff
@@ -0,0 +1,76 @@
+        status = str(exc)
+        raise self.retry()
+    finally:
+        log_message += status
```
Interesting that you define `log_message` way in the top and use it only here.

amatai replied:
Ah, a side effect of refactoring this code, I had logic to build the log message for different exceptions, later realized didn't need to do it that way.

anton-efimenko commented on minimailgun/store.py:
```diff
@@ -0,0 +1,118 @@
+
```
I would add some documentation instead of an empty line :)

amatai replied:
That is an area, I could have done much better overall in this PR. Agreed.

r0mant commented on minimailgun/api.py:
```diff
@@ -0,0 +1,96 @@
+
+@mailgun_api.route('/mail', methods=['POST'])
+def create_mail():
+    if request.content_type != 'application/json':
```
These two checks look generic enough so they may be useful across all handlers (routes) of the application, if we want to make the API consistent (which we should).

How about we create our own decorator that wraps `mailgun_api.route` and adds generic things that apply to all API methods (like these checks)? This will also free handlers from the noise, make them easier to read/understand.

amatai replied:
Yes, this is something I wanted to attack.

amatai replied:
Added a decorator and moved validation code out.

anton-efimenko commented on minimailgun/tests/base.py:
```diff
@@ -0,0 +1,10 @@
+
```
Have you by any chance tried to run the tests with code coverage?

amatai replied:
No, I didn't have much time to do that.

amatai replied:
```
coverage report | grep mini
minimailgun/__init__.py                                                                                               0      0   100%
minimailgun/api.py                                                                                                   56     16    71%
minimailgun/config.py                                                                                                 4      0   100%
minimailgun/store.py                                                                                                 70     11    84%
minimailgun/tasks.py                                                                                                 45     29    36%
minimailgun/tests/__init__.py                                                                                         0      0   100%
minimailgun/tests/base.py                                                                                             3      0   100%
minimailgun/tests/test_api.py                                                                                        31      0   100%
minimailgun/tests/test_store.py                                                                                      35      0   100%
```

anton-efimenko commented on minimailgun/api.py:
```diff
@@ -0,0 +1,96 @@
+    request_data = request.get_json()
+    if not request_data.get('from'):
+        abort(400, 'The from address is required.')
+    # other payload validation goes here.
```
Not sure about this comment.

amatai replied:
One thing I wanted to add was regex validation of email addresses being provided in from, to, cc etc.

anton-efimenko commented on minimailgun/store.py:
```diff
@@ -0,0 +1,118 @@
+    def update_status(self, id, rcpt, status):
+        rcpt = self._sanitize_for_bson(rcpt)
+        update_key = '_recipients.' + rcpt + '.'
+        self.db.mails.find_and_modify(
```
Is there any particular reason you use `find_and_modify` here instead of `update`?

amatai replied:
This came from my previous understanding that find_and_modify is atomic, which is what I wanted.

r0mant replied:
What about `update`? Isn't it atomic?

amatai replied:
As, I said, this came from my previous knowledge and my understanding from then. It seems update is atomic, now that I read the docs - but definitely not called out specifically on update documentation, rather find_and_modify states that update is atomic.
http://docs.mongodb.org/manual/reference/method/db.collection.update/
http://docs.mongodb.org/manual/reference/method/db.collection.findAndModify/#comparisons-with-the-update-method

amatai replied:
Just to clarify, when I had initially read about update vs find_and_modify, my understanding what `update` isn't atomic - maybe I was wrong about it from the beginning. Hence, this time, I went with that my existing understanding and used `find_and_modify`

amatai replied:
Changed that to `update` instead of `find_and_modify` since the place find_and_modify shines is if we need to fetch a document after update, and we don't use this functionality.

anton-efimenko commented on minimailgun/store.py:
```diff
@@ -0,0 +1,118 @@
+    def add_mail(self, mail, recipient_set):
+        mail['_created_at'] = datetime.utcnow()
+        # create a recipient list and adjust the key, as bson docs
+        # don't allow key with a '.' in it.
```
As an idea you can use separate documents for each recipient if you want to avoid sanitization.

r0mant replied:
+1

I'm too not very fond of idea of using recipient emails as keys. Email addresses can contain much more characters that could cause problems in addition to dots.

amatai replied:
Hmm, okay - didn't know about additional characters. Initially, I was trying to keep individual documents for each recipient and even trying to keep status from all 3 tries, so could have historical status in the db and not just logs. It was getting complex and in an effort to keep this simple and admittedly, also because to time (still have to do the day job) - I did it this way.

anton-efimenko replied:
No need for historical tracking. For simplicity you can give each recipient a unique ID even if they were submitted from a single API request.

amatai replied:
Changed the code so that email address is no longer used as the key for dictionary. Instead using a uuid. The message is still kept as an aggregate of all the recipients, because, want to maintain a link between the main message and its recipients.
We can certainly, create an entry per recipient, though that will require - establishing a link between the recipient document and the main message document (foreign-key relationship, in SQL terminology).

r0mant commented on minimailgun/tasks.py:
```diff
@@ -0,0 +1,76 @@
+    default_retry_delay=config['mail_properties']['retry_time'],
+    max_retries=config['mail_properties']['max_retry']
+)
+def sendmail(self, message, rcpt):
```
This function seems to be doing too much. Querying MX records and sending a mail are two separate tasks. It will make the logic clearer and the code generally more testable.

As a disclaimer, we usually try to adhere to Unix philosophy when writing code. Each function should preferably do just one thing. Without extremes, of course, but in this case clearly two independent things are going on.

amatai replied:
Broke down the method to two independent tasks.

r0mant commented on minimailgun/store.py:
```diff
@@ -0,0 +1,118 @@
+        self._db = None
+
+    def setup(self):
+        # TODO: setup indexing here
```
What about indexes?

amatai replied:
Initially, I thought I will need indexes for querying the data. But then it turned out that all my queries were based on mail-id, which is _id in the mongo doc and automatically indexed. Hence, didn't need any more indexes. If we change the model, this will be helper method to call and ensure indexes exist at startup.

anton-efimenko replied:
Makes sense, but you will be able to add this function in the future if needed.

amatai replied:
Yes, but I'm also trying to demonstrate my thinking/approach in this PR.

r0mant commented on minimailgun/api.py:
```diff
@@ -0,0 +1,96 @@
+
+
+if __name__ == '__main__':
+    app.run()
```
Can this app handle more than 1 client simultaneously?

amatai replied:
Sure, if run under a wsgi container like gunicorn, uwsgi - am I missing something with respect to  multiple clients?

r0mant commented on minimailgun/store.py:
```diff
@@ -0,0 +1,118 @@
+
+    @property
+    def client(self):
+        if not self._client:
```
You're doing lazy initialization here. What will happen if multiple threads try to access this property at the same time?

amatai replied:
Hmm, true in case of multi-threaded deploy, this could overwrite the client handle. In all honesty, in this particular case, it is not catastrophic, as the value overwritten will be the same always. Well, in that case, I don't see a reason to add the overhead of synchronization.

r0mant commented on minimailgun/api.py:
```diff
@@ -0,0 +1,96 @@
+        debug = bool(os.environ.get('MINI_MAILGUN_DEBUG'))
+    app.debug = debug
+    app.config.update(JSONIFY_PRETTYPRINT_REGULAR=JSONIFY_PRETTYPRINT_REGULAR, JSON_AS_ASCII=JSON_AS_ASCII)
+    for code in default_exceptions.keys():
```
What is `default_exceptions`?

amatai replied:
It is a dictionary of http error codes exceptions associated with them. The motivation for this code came from: http://flask.pocoo.org/snippets/83/

Commit:
Add docs and move code around - code review comments.

r0mant commented on minimailgun/store.py:
```diff
@@ -0,0 +1,118 @@
+            raise UnablToAddMessageError('Multiple tries led to DuplicateKeyError. Unable to add message id: {id}'.format(
+                id=mail['_id']
+            ))
+        return self.get_mail_by_id(result)
```
Why do you need to query for the object you've just inserted? If you want to return it, you can return the object you already have, can't you?

amatai replied:
Yes. In this case, we can.

Commit:
Return the existing doc, no need to fetch from db

Commit:
Remove empty setup method.

Commit:
Break down method to single responsibilities

Commit:
Move api validation code to a decorator

Commit:
find_and_modify -> update

Both are atomic, it is fetch after update is where find_and_modify
shines. We don't need that functionality.
Also fixed broken unit test.

Commit:
Changed recipient keys from email to uuid

Changed how the documents are stored, instead of using sanitized version
on email addresses, used a uuid now.

r0mant commented on minimailgun/api.py:
```diff
@@ -0,0 +1,118 @@
+from minimailgun.tasks import handle_new_message
+
+
+class validate_mimetype(object):
```
When I spoke about the decorator, I meant the decorator that would combine the validation with configuring a route so you'd have just a single "custom-made" decorator that you could apply to all handlers.

``` python
def api(path, methods, ...):
    # decorator that does validation and registers a route
    ...

@api("/mail", methods=["POST"])
def create_mail():
    ...
```

That is what we have in our API, but what you implemented also does the job. JFYI.

amatai replied:
Ah. Makes sense, thanks for sharing.

r0mant commented on minimailgun/store.py:
```diff
@@ -0,0 +1,118 @@
+        # its going to be rare that we'll have UUID collision
+        # its going to be extremely rare that two consecutive collisions
+        result = None
+        for _ in range(2):
```
I side with Anton on this one. Paranoia is good sometimes but not to extreme.

amatai replied:
Yes, I agree that it was probably excessive. But, it still worries me, what if there is a collision, what's the right behavior - just throw a 500 back and have the client retry, or try to salvage the request when we know the exact reason for failure.

amatai commented:
@r0mant @anton-efimenko I have addressed the concerns and answered the questions in this PR - let me know if you've had chance to review those and if you are further comments/questions.

r0mant commented:
@amatai lgtm
